### PR TITLE
cmake: use or turn off liburing for rocksdb

### PR DESCRIPTION
**Pull request by batrick**
There are two problems:

1. The librocksdb does not use the built liburing if present.
2. You cannot turn off rocksdb's use of liburing with WITH_LIBURING=OFF.

This commit resolves (2) and partially resolves (1). The build of targets still wrongly orders liburing.a before librocksdb.a so we continue to get errors like the one below.

This is an attempt to fix errors like:

	FAILED: bin/ceph_test_keyvaluedb_iterators
	: && /opt/rh/gcc-toolset-11/root/usr/bin/g++ -Og -g -rdynamic -pie src/test/ObjectMap/CMakeFiles/ceph_test_keyvaluedb_iterators.dir/test_keyvaluedb_iterators.cc.o src/test/ObjectMap/CMakeFiles/ceph_test_keyvaluedb_iterators.dir/KeyValueDBMemory.cc.o -o bin/ceph_test_keyvaluedb_iterators  -Wl,-rpath,/home/pdonnell/scratch/build/lib  lib/libos.a  lib/libgmock_maind.a  lib/libgmockd.	a  lib/libgtestd.a  -lpthread  -ldl  lib/libglobal.a  -ldl  /usr/lib64/librt.so  -lresolv  -ldl  lib/libblk.a  /lib64/libaio.so  src/liburing/src/liburing.a  lib/libkv.a  lib/libheap_profiler.a  /lib64/libtcmalloc.so  src/rocksdb/librocksdb.a  /lib64/libsnappy.so  /usr/lib64/liblz4.so  /usr/lib64/libz.so  /usr/lib64/libfuse.so  lib/libceph-common.so.2  src/opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a  src/opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a  src/opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a  src/opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a  src/opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a  /usr/lib64/libcurl.so  /usr/lib64/libthrift.so  lib/libjson_spirit.a  lib/libcommon_utf8.a  lib/liberasure_code.a  lib/libextblkdev.a  -lcap  boost/lib/libboost_thread.a  boost/lib/libboost_chrono.a  boost/lib/libboost_atomic.a  boost/lib/libboost_system.a  boost/lib/libboost_random.a  boost/lib/libboost_program_options.a  boost/lib/libboost_date_time.a  boost/lib/libboost_iostreams.a  boost/lib/libboost_regex.a  lib/libfmtd.a  /usr/lib64/libblkid.so  -lpthread  /usr/lib64/libcrypto.so  /usr/lib64/libudev.so  /usr/lib64/libz.so  -ldl  -lresolv   -Wl,--as-needed -latomic && :
	/opt/rh/gcc-toolset-11/root/usr/bin/ld: src/rocksdb/librocksdb.a(fs_posix.cc.o): in function `io_uring_wait_cqe_nr':
	/home/pdonnell/scratch/build/src/liburing/src/include/liburing.h:494: undefined reference to `__io_uring_get_cqe'
	/opt/rh/gcc-toolset-11/root/usr/bin/ld: src/rocksdb/librocksdb.a(fs_posix.cc.o): in function `rocksdb::(anonymous namespace)::PosixFileSystem::AbortIO(std::vector<void*, std::allocator<void*> >&)':
	/home/pdonnell/ceph/src/rocksdb/env/fs_posix.cc:1125: undefined reference to `io_uring_get_sqe'
	/opt/rh/gcc-toolset-11/root/usr/bin/ld: /home/pdonnell/ceph/src/rocksdb/env/fs_posix.cc:1134: undefined reference to `io_uring_submit'
	/opt/rh/gcc-toolset-11/root/usr/bin/ld: src/rocksdb/librocksdb.a(fs_posix.cc.o): in function `rocksdb::CreateIOUring()':
	/home/pdonnell/ceph/src/rocksdb/env/io_posix.h:272: undefined reference to `io_uring_queue_init'
	/opt/rh/gcc-toolset-11/root/usr/bin/ld: src/rocksdb/librocksdb.a(io_posix.cc.o): in function `io_uring_wait_cqe_nr':
	/home/pdonnell/scratch/build/src/liburing/src/include/liburing.h:494: undefined reference to `__io_uring_get_cqe'
	/opt/rh/gcc-toolset-11/root/usr/bin/ld: src/rocksdb/librocksdb.a(io_posix.cc.o): in function `rocksdb::PosixRandomAccessFile::MultiRead(rocksdb::FSReadRequest*, unsigned long, rocksdb::IOOptions const&, rocksdb::IODebugContext*)':
	/home/pdonnell/ceph/src/rocksdb/env/io_posix.cc:674: undefined reference to `io_uring_get_sqe'
	/opt/rh/gcc-toolset-11/root/usr/bin/ld: /home/pdonnell/ceph/src/rocksdb/env/io_posix.cc:684: undefined reference to `io_uring_submit_and_wait'
	/opt/rh/gcc-toolset-11/root/usr/bin/ld: src/rocksdb/librocksdb.a(io_posix.cc.o): in function `rocksdb::PosixRandomAccessFile::ReadAsync(rocksdb::FSReadRequest&, rocksdb::IOOptions const&, std::function<void (rocksdb::FSReadRequest const&, void*)>, void*, void**, std::function<void (void*)>*, rocksdb::IODebugContext*)':
	/home/pdonnell/ceph/src/rocksdb/env/io_posix.cc:901: undefined reference to `io_uring_get_sqe'
	/opt/rh/gcc-toolset-11/root/usr/bin/ld: /home/pdonnell/ceph/src/rocksdb/env/io_posix.cc:910: undefined reference to `io_uring_submit'
	collect2: error: ld returned 1 exit status

Fixes: https://tracker.ceph.com/issues/63218

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: cmake/modules/BuildRocksDB.cmake [hunk 1]
```diff
@@ -86,6 +93,15 @@ function(build_rocksdb)
...
   add_library(RocksDB::RocksDB STATIC IMPORTED)
   add_dependencies(RocksDB::RocksDB rocksdb_ext)
+  if(WITH_LIBURING)
+    # None of these work!
+    #add_dependencies(RocksDB::RocksDB INTERFACE uring::uring)
+    #add_dependencies(rocksdb_ext uring::uring)
+    #target_link_libraries(rocksdb_ext PRIVATE uring::uring)
+    #target_link_libraries(RocksDB::RocksDB PRIVATE uring::uring)
+    #target_link_libraries(RocksDB::RocksDB INTERFACE uring::uring)
+    #add_dependencies(RocksDB::RocksDB uring::uring)
```
Comment (batrick) [inline cmake/modules/BuildRocksDB.cmake:103]: Unfortunately, none of these commands work to correctly order the dependency of `src/rocksdb/librocksdb.a` on `src/liburing/src/liburing.a`. I've done a lot of googling and experimentation without success to get this to work.

Comment (batrick) [reply cmake/modules/BuildRocksDB.cmake:103]: cc @cbodley @rzarzynski 

Any other cmake experts that may have an idea?

Comment (cbodley) [reply cmake/modules/BuildRocksDB.cmake:103]: other rocksdb dependencies are added above with the form:
```cmake
  if(SNAPPY_FOUND)
    list(APPEND rocksdb_INTERFACE_LINK_LIBRARIES snappy::snappy)
  endif()
```
then applied at the end:
```cmake
  set_target_properties(RocksDB::RocksDB PROPERTIES
    INTERFACE_INCLUDE_DIRECTORIES "${rocksdb_INCLUDE_DIR}"
    INTERFACE_LINK_LIBRARIES "${rocksdb_INTERFACE_LINK_LIBRARIES}"
```
i'd expect that use of `INTERFACE_LINK_LIBRARIES` to be equivalent to your line:
```cmake
target_link_libraries(RocksDB::RocksDB INTERFACE uring::uring)
```
is it possible that `set_target_properties()` is overwriting that?

Comment (batrick) [reply cmake/modules/BuildRocksDB.cmake]: > other rocksdb dependencies are added above with the form:
> 
> ```cmake
>   if(SNAPPY_FOUND)
>     list(APPEND rocksdb_INTERFACE_LINK_LIBRARIES snappy::snappy)
>   endif()
> ```

I can't believe I missed this. That was all we needed!

**Review by cbodley — APPROVED** [review]
looks good if it works 👍

File: cmake/modules/BuildRocksDB.cmake (hunk 1 above)
Comment (tchaikov) [reply cmake/modules/BuildRocksDB.cmake]: sorry, i missed this. i started commenting right after reading "correctly order the dependency of".

> list(APPEND rocksdb_CMAKE_ARGS -DWITH_GFLAGS=OFF)

could you help me understand how this is related to liburing?

also, i'd suggest to split this change into two. as they addresses two different issues:

1. promote the uring to top source directory
2. populate liburing include and library paths down to rocksdb external project

Comment (batrick) [reply cmake/modules/BuildRocksDB.cmake]: I don't recall why I turned GFLAGS off. I reverted that and applied your other suggestions. Thanks @tchaikov

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#16-Oct-2023

**Comment by cyx1231st** [comment]
I met a build error which looks related to iouring, and identified that it is somehow related to this PR. (It starts to happen right after this PR).

I start with a clean Ubuntu 22.04.1 LTS and do the following steps as root, e.g.:
```
$ cd ceph
$ WITH_SEASTAR=1 NO_BOOST_PKGS=1 ./install-deps.sh
$ cd build
$ cmake -DWITH_CCACHE=ON -DWITH_MGR=OFF -DWITH_SPDK=OFF -DWITH_CEPHFS=OFF -DWITH_RADOSGW=OFF -DWITH_FUSE=OFF -DWITH_MGR_DASHBOARD_FRONTEND=OFF -DWITH_KRBD=OFF -DWITH_LEVELDB=OFF -DWITH_JAEGER=OFF -DWITH_TESTS=OFF -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_CXX_FLAGS="-fno-omit-frame-pointer" -DWITH_SEASTAR=ON -GNinja -DWITH_PYTHON3=3 ..
$ ninja vstart crimson-osd rbd
```

I'm not sure what happened yet, looks the direct error message is:
```
CMake Error in CMakeLists.txt:
  Imported target "uring::uring" includes non-existent path

    "/root/ceph_builds/obj-x86_64-linux-gnu-info/src/liburing/src/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

And the complete log is at: https://gist.github.com/cyx1231st/6630e42d19bd0da5dabe568526476c23

**Comment by cbodley** [comment]
> And the complete log is at: https://gist.github.com/cyx1231st/6630e42d19bd0da5dabe568526476c23

@cyx1231st from the log, it looks like the rocksdb build failed before the uring build started, so rocksdb wasn't able to find its include directory. i opened https://github.com/ceph/ceph/pull/54626 that should hopefully fix the ordering

**Pull request merged**